File: services/client/src/protocol/protocol.go
package protocol

import (
	"bytes"
	"encoding/binary"
	"fmt"
	"net"

	"github.com/7574-sistemas-distribuidos/tp-nivelador/src/domain"
	"github.com/7574-sistemas-distribuidos/tp-nivelador/src/safe_socket"
)

const (
	MsgBet     byte = 0x01 // Apuesta
	MsgEnd     byte = 0x02 // Fin de apuestas
	MsgWinners byte = 0x03 // Respuesta del servidor con lista de ganadores
	MsgAck     byte = 0x04 // Ack del server tras recibir un batch de apuestas
)

const HeaderSize = 5 // 1 byte para el tipo, 4 para el largo del mensaje
const AgencySize = 4
const DocSize = 4
const BirthSize = 10
const NumSize = 4

func SerializeBet(agencyID uint32, bet *domain.Bet) []byte {
	firstNameBytes := []byte(bet.FirstName)
	lastNameBytes := []byte(bet.LastName)
	birthdateBytes := []byte(bet.Birthdate)

	payloadSize := AgencySize + 1 + len(firstNameBytes) + 1 + len(lastNameBytes) + DocSize + BirthSize + NumSize
	payload := make([]byte, payloadSize)

	offset := 0

	binary.BigEndian.PutUint32(payload[offset:offset+AgencySize], agencyID)
	offset += AgencySize

	payload[offset] = byte(len(firstNameBytes))
	offset++
	copy(payload[offset:], firstNameBytes)
	offset += len(firstNameBytes)

	payload[offset] = byte(len(lastNameBytes))
	offset++
	copy(payload[offset:], lastNameBytes)
	offset += len(lastNameBytes)

	binary.BigEndian.PutUint32(payload[offset:offset+DocSize], uint32(bet.Document))
	offset += DocSize

	copy(payload[offset:], birthdateBytes)
	offset += BirthSize

	binary.BigEndian.PutUint32(payload[offset:offset+NumSize], uint32(bet.Number))

	return payload
}

func SendBetBatch(conn net.Conn, agencyID uint32, bets []*domain.Bet) error {
	var payloadBuffer bytes.Buffer

	for _, bet := range bets {
		payloadBuffer.Write(SerializeBet(agencyID, bet))
	}

	payload := payloadBuffer.Bytes()
	payloadLen := uint32(len(payload))

	var packetBuffer bytes.Buffer

	batchHeader := make([]byte, HeaderSize)
	batchHeader[0] = MsgBet
	binary.BigEndian.PutUint32(batchHeader[1:5], payloadLen)

	packetBuffer.Write(batchHeader)
	packetBuffer.Write(payload)

	return safe_socket.SendAll(conn, packetBuffer.Bytes())
}

func WaitBatchAck(conn net.Conn) error {
	header, err := safe_socket.RecvAll(conn, HeaderSize)
	if err != nil {
		return fmt.Errorf("error leyendo header de ack: %w", err)
	}

	msgType := header[0]
	if msgType != MsgAck {
		return fmt.Errorf("tipo de mensaje inesperado: %d", msgType)
	}

	return nil
}

func SendEnd(conn net.Conn) error {
	header := make([]byte, HeaderSize)
	header[0] = MsgEnd
	binary.BigEndian.PutUint32(header[1:5], 0)

	return safe_socket.SendAll(conn, header)
}

func ReceiveWinners(conn net.Conn) ([]string, error) {
	header, err := safe_socket.RecvAll(conn, HeaderSize)
	if err != nil {
		return nil, fmt.Errorf("error leyendo header de respuesta: %w", err)
	}

	msgType := header[0]
	if msgType != MsgWinners {
		return nil, fmt.Errorf("tipo de mensaje inesperado: %d", msgType)
	}

	payloadLen := binary.BigEndian.Uint32(header[1:5])
	if payloadLen == 0 {
		return []string{}, nil
	}

	payload, err := safe_socket.RecvAll(conn, int(payloadLen))
	if err != nil {
		return nil, fmt.Errorf("error leyendo ganadores: %w", err)
	}

	winners := []string{}
	offset := 0
	totalLen := len(payload)

	for offset < totalLen {
		fnLen := int(payload[offset])
		offset++
		firstName := string(payload[offset : offset+fnLen])
		offset += fnLen

		lnLen := int(payload[offset])
		offset++
		lastName := string(payload[offset : offset+lnLen])
		offset += lnLen

		document := binary.BigEndian.Uint32(payload[offset : offset+DocSize])
		offset += DocSize

		birthdate := string(payload[offset : offset+BirthSize])
		offset += BirthSize

		number := binary.BigEndian.Uint32(payload[offset : offset+NumSize])
		offset += NumSize

		winnerStr := fmt.Sprintf("%s,%s,%d,%s,%d", firstName, lastName, document, birthdate, number)
		winners = append(winners, winnerStr)
	}

	return winners, nil
}

File: services/server/src/protocol/protocol.py
import lottery

import logger

MSG_BET = 1
MSG_END = 2
MSG_WINNERS = 3
MSG_ACK = 4

HEADER_SIZE = 5 # 1 byte para el tipo, 4 para el largo del mensaje
AGENCY_SIZE = 4
DOC_SIZE = 4
BIRTH_SIZE = 10
NUM_SIZE = 4

def deserialize_bet_batch(payload: bytes) -> list[lottery.Bet]:
    bets = []
    offset = 0
    total_len = len(payload)

    while offset < total_len:
        agency_id = int.from_bytes(payload[offset:offset+AGENCY_SIZE], byteorder='big')
        offset += AGENCY_SIZE
    
        fn_len = payload[offset]
        offset += 1
        first_name = payload[offset:offset+fn_len].decode('utf-8')
        offset += fn_len
    
        ln_len = payload[offset]
        offset += 1
        last_name = payload[offset:offset+ln_len].decode('utf-8')
        offset += ln_len
    
        document = int.from_bytes(payload[offset:offset+DOC_SIZE], byteorder='big')
        offset += DOC_SIZE
    
        birthdate = payload[offset:offset+BIRTH_SIZE].decode('utf-8')
        offset += BIRTH_SIZE
    
        number = int.from_bytes(payload[offset:offset+NUM_SIZE], byteorder='big')
        offset += NUM_SIZE
        

        bets.append(lottery.Bet(
            agency_id=agency_id,
            first_name=first_name,
            last_name=last_name,
            document=document,
            birthdate=birthdate,
            number=number,
        ))

    return bets

def serialize_ack() -> bytes:
    header = bytearray(HEADER_SIZE)
    header[0] = MSG_ACK
    header[1:5] = (0).to_bytes(4, byteorder='big')
    return bytes(header)

def serialize_winners(winners: list[lottery.Bet]) -> bytes:
    payload = bytearray()

    for bet in winners:
        fn_bytes = bet.first_name.encode('utf-8')
        payload.append(len(fn_bytes))
        payload.extend(fn_bytes)

        ln_bytes = bet.last_name.encode('utf-8')
        payload.append(len(ln_bytes))
        payload.extend(ln_bytes)

        payload.extend(bet.document.to_bytes(DOC_SIZE, byteorder='big'))

        payload.extend(bet.birthdate.encode('utf-8'))

        payload.extend(bet.number.to_bytes(NUM_SIZE, byteorder='big'))

    payload_len = len(payload)

    header = bytearray(HEADER_SIZE)
    header[0] = MSG_WINNERS
    header[1:5] = payload_len.to_bytes(4, byteorder='big')

    return bytes(header + payload)